File: backend/game/rules.py
from dataclasses import dataclass


DIRECTIONS = ((1, 0), (0, 1), (1, 1), (1, -1))


@dataclass
class RuleResult:
    ok: bool
    reason: str = ""
    status: str = "playing"
    winner: str = ""
    forbidden: bool = False


def opponent(color):
    return "white" if color == "black" else "black"


def key(x, y):
    return f"{x},{y}"


def board_from(stones):
    return {key(stone["x"], stone["y"]): stone["color"] for stone in stones}


def inside(x, y, board_size=15):
    return 0 <= x < board_size and 0 <= y < board_size


def count_dir(board, x, y, dx, dy, color):
    count = 0
    cx = x + dx
    cy = y + dy
    while board.get(key(cx, cy)) == color:
        count += 1
        cx += dx
        cy += dy
    return count


def line_cells(board, x, y, dx, dy, board_size=15):
    cells = []
    for offset in range(-5, 6):
        cx = x + dx * offset
        cy = y + dy * offset
        value = board.get(key(cx, cy))
        if not inside(cx, cy, board_size):
            cells.append("X")
        elif value is None:
            cells.append(".")
        else:
            cells.append("B" if value == "black" else "W")
    return cells, 5


def window_includes_center(start, length, center):
    return start <= center < start + length


def line_positions(x, y, dx, dy, board_size=15):
    sx, sy = x, y
    while inside(sx - dx, sy - dy, board_size):
        sx -= dx
        sy -= dy
    positions = []
    while inside(sx, sy, board_size):
        positions.append((sx, sy))
        sx += dx
        sy += dy
    return positions


def open_four_exists(board, x, y, dx, dy, color, board_size=15, required=()):
    required = set(required)
    positions = line_positions(x, y, dx, dy, board_size)
    values = [board.get(key(px, py), ".") for px, py in positions]
    for start in range(0, len(values) - 5):
        window = values[start : start + 6]
        if window != [".", color, color, color, color, "."]:
            continue
        window_positions = set(positions[start : start + 6])
        if required.issubset(window_positions):
            return True
    return False


def exact_five_exists(board, x, y, dx, dy, color, board_size=15, required=()):
    required = set(required)
    positions = line_positions(x, y, dx, dy, board_size)
    values = [board.get(key(px, py), ".") for px, py in positions]
    for start in range(0, len(values) - 4):
        window = values[start : start + 5]
        if window != [color, color, color, color, color]:
            continue
        before = values[start - 1] if start > 0 else "."
        after = values[start + 5] if start + 5 < len(values) else "."
        if before == color or after == color:
            continue
        window_positions = set(positions[start : start + 5])
        if required.issubset(window_positions):
            return True
    return False


def winning_extensions_in_direction(board, x, y, dx, dy, color, board_size=15, required=()):
    required = set(required) | {(x, y)}
    wins = []
    for px, py in line_positions(x, y, dx, dy, board_size):
        if board.get(key(px, py)) is not None:
            continue
        test_board = {**board, key(px, py): color}
        if exact_five_exists(test_board, x, y, dx, dy, color, board_size, required=required | {(px, py)}):
            wins.append((px, py))
    return wins


def direction_has_four(board, x, y, dx, dy, color, board_size=15):
    return bool(winning_extensions_in_direction(board, x, y, dx, dy, color, board_size))


def pattern_open_three_exists(board, x, y, dx, dy, color, board_size=15):
    patterns = (
        (".", color, color, color, "."),
        (".", color, color, ".", color, "."),
        (".", color, ".", color, color, "."),
        (".", color, ".", color, ".", color, "."),
    )
    positions = line_positions(x, y, dx, dy, board_size)
    values = [board.get(key(px, py), ".") for px, py in positions]
    for pattern in patterns:
        size = len(pattern)
        for start in range(0, len(values) - size + 1):
            if tuple(values[start : start + size]) != pattern:
                continue
            if (x, y) in set(positions[start : start + size]):
                return True
    return False


def direction_has_open_three(board, x, y, dx, dy, color, board_size=15):
    if pattern_open_three_exists(board, x, y, dx, dy, color, board_size):
        return True
    for px, py in line_positions(x, y, dx, dy, board_size):
        if board.get(key(px, py)) is not None:
            continue
        test_board = {**board, key(px, py): color}
        if open_four_exists(test_board, x, y, dx, dy, color, board_size, required=((x, y), (px, py))):
            return True
        if len(winning_extensions_in_direction(test_board, x, y, dx, dy, color, board_size, required=((x, y), (px, py)))) >= 2:
            return True
    return False


def validate_move(stones, x, y, color, rule_set="standard", board_size=15):
    board = board_from(stones)
    if not inside(x, y, board_size):
        return RuleResult(False, "落子超出棋盘范围")
    if key(x, y) in board:
        return RuleResult(False, "该交点已有棋子")
    if rule_set != "renju" or color != "black":
        return RuleResult(True)

    next_board = board_from([*stones, {"x": x, "y": y, "color": color}])
    line_lengths = [
        1 + count_dir(next_board, x, y, dx, dy, color) + count_dir(next_board, x, y, -dx, -dy, color)
        for dx, dy in DIRECTIONS
    ]
    if any(length > 5 for length in line_lengths):
        return RuleResult(True, "黑棋长连禁手，白棋获胜", status="white_win", winner="white", forbidden=True)
    if any(length == 5 for length in line_lengths):
        return RuleResult(True)

    four_count = 0
    three_count = 0
    for dx, dy in DIRECTIONS:
        four_count += int(direction_has_four(next_board, x, y, dx, dy, color, board_size))
        three_count += int(direction_has_open_three(next_board, x, y, dx, dy, color, board_size))
    if four_count >= 2:
        return RuleResult(True, "黑棋双四禁手，白棋获胜", status="white_win", winner="white", forbidden=True)
    if three_count >= 2:
        return RuleResult(True, "黑棋双三禁手，白棋获胜", status="white_win", winner="white", forbidden=True)
    return RuleResult(True)


def evaluate_move(stones, x, y, color, rule_set="standard", board_size=15):
    validation = validate_move(stones, x, y, color, rule_set, board_size)
    if not validation.ok:
        return validation
    if validation.forbidden:
        return validation

    next_stones = [*stones, {"x": x, "y": y, "color": color}]
    board = board_from(next_stones)
    max_line = max(
        1 + count_dir(board, x, y, dx, dy, color) + count_dir(board, x, y, -dx, -dy, color)
        for dx, dy in DIRECTIONS
    )
    wins = max_line == 5 if rule_set == "renju" and color == "black" else max_line >= 5
    if wins:
        return RuleResult(True, status=f"{color}_win", winner=color)
    if len(next_stones) >= board_size * board_size:
        return RuleResult(True, status="draw")
    return RuleResult(True)


def next_turn(stones):
    return "black" if len(stones) % 2 == 0 else "white"

File: frontend/src/rules.ts
import type { GameStatus, Move, RuleSet, StoneColor } from "./types";

interface RuleCheck {
  ok: boolean;
  reason?: string;
  status?: GameStatus;
  winner?: StoneColor;
  forbidden?: boolean;
}

const directions = [
  [1, 0],
  [0, 1],
  [1, 1],
  [1, -1],
] as const;

const key = (x: number, y: number) => `${x},${y}`;
const inside = (x: number, y: number, size = 15) => x >= 0 && y >= 0 && x < size && y < size;

function boardFrom(stones: Move[]) {
  return new Map(stones.map((stone) => [key(stone.x, stone.y), stone.color]));
}

function count(board: Map<string, StoneColor>, x: number, y: number, dx: number, dy: number, color: StoneColor) {
  let total = 0;
  let cx = x + dx;
  let cy = y + dy;
  while (board.get(key(cx, cy)) === color) {
    total += 1;
    cx += dx;
    cy += dy;
  }
  return total;
}

function linePositions(x: number, y: number, dx: number, dy: number, size: number) {
  let sx = x;
  let sy = y;
  while (inside(sx - dx, sy - dy, size)) {
    sx -= dx;
    sy -= dy;
  }
  const positions: Array<[number, number]> = [];
  while (inside(sx, sy, size)) {
    positions.push([sx, sy]);
    sx += dx;
    sy += dy;
  }
  return positions;
}

function hasAll(required: Array<[number, number]>, window: Set<string>) {
  return required.every(([x, y]) => window.has(key(x, y)));
}

function openFourExists(
  board: Map<string, StoneColor>,
  x: number,
  y: number,
  dx: number,
  dy: number,
  color: StoneColor,
  size: number,
  required: Array<[number, number]>,
) {
  const positions = linePositions(x, y, dx, dy, size);
  const values = positions.map(([px, py]) => board.get(key(px, py)) || ".");
  for (let i = 0; i <= values.length - 6; i += 1) {
    const window = values.slice(i, i + 6);
    if (window.join("") !== `.${color}${color}${color}${color}.`) continue;
    if (hasAll(required, new Set(positions.slice(i, i + 6).map(([px, py]) => key(px, py))))) return true;
  }
  return false;
}

function exactFiveExists(
  board: Map<string, StoneColor>,
  x: number,
  y: number,
  dx: number,
  dy: number,
  color: StoneColor,
  size: number,
  required: Array<[number, number]>,
) {
  const positions = linePositions(x, y, dx, dy, size);
  const values = positions.map(([px, py]) => board.get(key(px, py)) || ".");
  const five = `${color}${color}${color}${color}${color}`;
  for (let i = 0; i <= values.length - 5; i += 1) {
    if (values.slice(i, i + 5).join("") !== five) continue;
    if (values[i - 1] === color || values[i + 5] === color) continue;
    if (hasAll(required, new Set(positions.slice(i, i + 5).map(([px, py]) => key(px, py))))) return true;
  }
  return false;
}

function winningExtensionsInDirection(
  board: Map<string, StoneColor>,
  x: number,
  y: number,
  dx: number,
  dy: number,
  color: StoneColor,
  size: number,
  required: Array<[number, number]> = [],
) {
  const wins: Array<[number, number]> = [];
  const baseRequired: Array<[number, number]> = [[x, y], ...required];
  for (const [px, py] of linePositions(x, y, dx, dy, size)) {
    if (board.has(key(px, py))) continue;
    const next = new Map(board);
    next.set(key(px, py), color);
    if (exactFiveExists(next, x, y, dx, dy, color, size, [...baseRequired, [px, py]])) wins.push([px, py]);
  }
  return wins;
}

function directionHasFour(board: Map<string, StoneColor>, x: number, y: number, dx: number, dy: number, color: StoneColor, size: number) {
  return winningExtensionsInDirection(board, x, y, dx, dy, color, size).length > 0;
}

function patternOpenThreeExists(board: Map<string, StoneColor>, x: number, y: number, dx: number, dy: number, color: StoneColor, size: number) {
  const patterns = [
    [".", color, color, color, "."],
    [".", color, color, ".", color, "."],
    [".", color, ".", color, color, "."],
    [".", color, ".", color, ".", color, "."],
  ];
  const positions = linePositions(x, y, dx, dy, size);
  const values = positions.map(([px, py]) => board.get(key(px, py)) || ".");
  for (const pattern of patterns) {
    for (let i = 0; i <= values.length - pattern.length; i += 1) {
      if (values.slice(i, i + pattern.length).every((value, index) => value === pattern[index])) {
        if (positions.slice(i, i + pattern.length).some(([px, py]) => px === x && py === y)) return true;
      }
    }
  }
  return false;
}

function directionHasOpenThree(board: Map<string, StoneColor>, x: number, y: number, dx: number, dy: number, color: StoneColor, size: number) {
  if (patternOpenThreeExists(board, x, y, dx, dy, color, size)) return true;
  for (const [px, py] of linePositions(x, y, dx, dy, size)) {
    if (board.has(key(px, py))) continue;
    const next = new Map(board);
    next.set(key(px, py), color);
    if (openFourExists(next, x, y, dx, dy, color, size, [[x, y], [px, py]])) return true;
    if (winningExtensionsInDirection(next, x, y, dx, dy, color, size, [[px, py]]).length >= 2) return true;
  }
  return false;
}

export function nextTurn(stones: Move[]): StoneColor {
  return stones.length % 2 === 0 ? "black" : "white";
}

export function opponent(color: StoneColor): StoneColor {
  return color === "black" ? "white" : "black";
}

export function validateMove(stones: Move[], x: number, y: number, color: StoneColor, ruleSet: RuleSet, size = 15): RuleCheck {
  const board = boardFrom(stones);
  if (!inside(x, y, size)) return { ok: false, reason: "落子超出棋盘范围" };
  if (board.has(key(x, y))) return { ok: false, reason: "该交点已有棋子" };
  if (ruleSet !== "renju" || color !== "black") return { ok: true };
  const nextBoard = boardFrom([...stones, { x, y, color }]);
  const lengths = directions.map(([dx, dy]) => 1 + count(nextBoard, x, y, dx, dy, color) + count(nextBoard, x, y, -dx, -dy, color));
  if (lengths.some((length) => length > 5)) return { ok: true, reason: "黑棋长连禁手，白棋获胜", status: "white_win", winner: "white", forbidden: true };
  if (lengths.some((length) => length === 5)) return { ok: true };
  let fours = 0;
  let threes = 0;
  for (const [dx, dy] of directions) {
    fours += directionHasFour(nextBoard, x, y, dx, dy, color, size) ? 1 : 0;
    threes += directionHasOpenThree(nextBoard, x, y, dx, dy, color, size) ? 1 : 0;
  }
  if (fours >= 2) return { ok: true, reason: "黑棋双四禁手，白棋获胜", status: "white_win", winner: "white", forbidden: true };
  if (threes >= 2) return { ok: true, reason: "黑棋双三禁手，白棋获胜", status: "white_win", winner: "white", forbidden: true };
  return { ok: true };
}

export function evaluateMove(stones: Move[], x: number, y: number, color: StoneColor, ruleSet: RuleSet, size = 15): RuleCheck & { status: GameStatus } {
  const validation = validateMove(stones, x, y, color, ruleSet, size);
  if (!validation.ok) return { ...validation, status: "playing" as GameStatus };
  if (validation.forbidden) return { ...validation, status: validation.status || "white_win" };
  const board = boardFrom([...stones, { x, y, color }]);
  const maxLine = Math.max(...directions.map(([dx, dy]) => 1 + count(board, x, y, dx, dy, color) + count(board, x, y, -dx, -dy, color)));
  const wins = ruleSet === "renju" && color === "black" ? maxLine === 5 : maxLine >= 5;
  if (wins) return { ok: true, status: `${color}_win` as GameStatus, winner: color };
  if (stones.length + 1 >= size * size) return { ok: true, status: "draw" as GameStatus };
  return { ok: true, status: "playing" as GameStatus };
}
